### PR TITLE
Feature/single server smart graph satellite

### PR DESCRIPTION
**Pull request by romanatarango**
### Scope & Purpose

Tests for dump and restore when dumping on single server and restoring in a cluster for SmartGraph, Disjoint SmartGraph, HybridSmartGraph and HybridDisjointSmartGraph. Added tests for empty graphs (without edge relations), graphs without data (with empty vertex and edge sets) and graphs with data:
- The smart graph is a cycle of length 100 with a symmetric edge relation and self-loops on every vertex.
- The disjoint graph is a disjoint union of two such cycles of size 50 each.
- The satellite graph is also a disjoint union of two such cycles of size 50 each.
- The hybrid graph has two vertex collections: one satellite and one smart each of size 50. In both vertex collections, the vertices are enumerated by `value` (which is the `smartGraphAttribute`) from 0 to 49. The edges go from vertex _i_ (in the satellite collection) to vertex _i_ (in the smart collection) and from vertex _i_ to vertex _(i+1) % 50_.
- The hybrid disjoint graph a graph with one satellite vertex collection (`verticesSat`), two non-satellite vertex collections (`verticesANonSat` and `verticesBNonSat`) connected to other collections, an orphan collection and two edge relations: `edgeSatA` from `verticesSat` to `verticesANonSat` and `edgesAB` from `verticesANonSat` to `verticesBNonSat`. The size of every vertex collection is 50. In each vertex collection, the vertices are enumerated by `value` (which is the `smartGraphAttribute`) from 0 to 49. Edges from `edgeSatA` go from vertex _i_ (in `verticesSat`) to vertex _i_ (in `verticesANonSat`) and from vertex _i_ to vertex _(i+1) % 50_. In total, we have 50 + 50 = 100 edges in this collection (this is tested). Edges from `edgeAB` are constructed with the same function but go only from vertex _i_ to vertex _i_ because both vertex collections are not satellite and vertices with different values of `value` (i.e., _i_ and _(i+1) % 50_) cannot be
connected in a disjoint graph. Thus in total we have 50 edges in this collection (this is tested).

The orphans are faked: just some vertex collections are used to fill them, but it is not tested whether the vertices are really orphans.

The tests verify that graph and collections properties of restored graphs are correct.

*(Please describe the changes in this PR for reviewers - **mandatory**)*

- [ ] :hankey: Bugfix (requires CHANGELOG entry)
- [ ] :pizza: New feature (requires CHANGELOG entry, feature documentation and release notes)
- [ ] :fire: Performance improvement
- [ ] :hammer: Refactoring/simplification
- [ ] :book: CHANGELOG entry made

#### Backports:

- [ ] No backports required
- [ ] Backports required for: *(Please specify versions and link PRs)*

#### Related Information

*(Please reference tickets / specification / other PRs etc)*

- [ ] Docs PR: 
- [ ] Enterprise PR:
- [ ] GitHub issue / Jira ticket number:
- [ ] Design document: 

### Testing & Verification

*(Please pick either of the following options)*

- [ ] This change is a trivial rework / code cleanup without any test coverage.
- [ ] The behavior in this PR was *manually tested*
- [ ] This change is already covered by existing tests, such as *(please describe tests)*.
- [x] This PR adds tests that were used to verify all changes:
  - [ ] Added new C++ **Unit tests**
  - [x] Added new **integration tests** (e.g. in shell_server / shell_server_aql)
  - [ ] Added new **resilience tests** (only if the feature is impacted by failovers)
- [ ] There are tests in an external testing repository:
- [ ] I ensured this code runs with ASan / TSan or other static verification tools

Link to Jenkins PR run:

### Documentation

> All new features should be accompanied by corresponding documentation. 
> Bugs and features should furthermore be documented in the CHANGELOG so that
> support, end users, and other developers have a concise overview. 

- [ ] Added entry to *Release Notes* 
- [ ] Added a new section in the *Manual* 
- [ ] Added a new section in the *HTTP API* 
- [ ] Added *Swagger examples* for the HTTP API  
- [ ] Updated license information in *LICENSES-OTHER-COMPONENTS.md* for 3rd party libraries

### External contributors / CLA Note 

Please note that for legal reasons we require you to sign the [Contributor Agreement](https://www.arangodb.com/documents/cla.pdf)
before we can accept your pull requests.

### CODE REVIEW

File: tests/js/server/dump/dump-setup-common.inc [hunk 1]
```diff
@@ -481,49 +562,338 @@ function createSmartGraphSingleServer() {
   const vertices = "UnitTestDumpSmartVerticesSingleServer";
   const orphans = "UnitTestDumpSmartOrphansSingleServer";
 
-  createSmartGraphWithoutData({ smartGraphName, edges, vertices, orphans });
+  createSmartGraphWithoutData({
```
[the remark this answers is not in the document]
Comment (romanatarango) [reply tests/js/server/dump/dump-setup-common.inc]: 1) changed to `createGraphWithoutData`
2) now exported as `createSmartGraphWithoutData`. Still confusing?

File: tests/js/server/dump/dump-setup-common.inc [hunk 2]
```diff
@@ -481,49 +562,338 @@ function createSmartGraphSingleServer() {
...
+
+  createSmartGraphWithoutData({
+      smartGraphName: graphName,
+      edges: edges,
+      fromVertices: [vertices],
+      toVertices: [vertices],
+      orphans: [orphans]
+    },
+    {
+      isSatellite: true, isSmart: false/*needed: default is true*/,
```
[the remark this answers is not in the document]
Comment (romanatarango) [reply tests/js/server/dump/dump-setup-common.inc:632]: Is there any case where we need `isSmart === false`? If not, `createGraphWithoutData` could remove the key `isSmart` if it obtains it with value `false`.

File: tests/js/server/dump/dump-setup-common.inc [hunk 3]
```diff
@@ -481,49 +562,338 @@ function createSmartGraphSingleServer() {
...
+  const edges = "UnitTestDumpHybridSmartEdgesSingleServer";
+  const orphans = "UnitTestDumpHybridSmartOrphansSingleServer";
+
+  if (sgm._exists(hybridGraphName)) {
+    sgm._drop(hybridGraphName, true);
+  }
+  db._drop(edges);
+  db._drop(verticesNonSat);
+  db._drop(verticesSat);
+  db._drop(orphans);
```
[the remark this answers is not in the document]
Comment (romanatarango) [reply tests/js/server/dump/dump-setup-common.inc:665]: refactored: new function assertCollectionsAreNew([...]) throws if one of given strings is the name of an existing collection

File: tests/js/server/dump/dump-test.inc [hunk 4]
```diff
@@ -1281,7 +1377,203 @@
...
+
+      const g = db._collection('_graphs');
+      assertTrue(!!g.exists(hybridGraphName));
+
+      const graph = g.document(hybridGraphName);
+      assertEqual(graph.isSmart, undefined);
+      assertFalse(graph.isSatellite);
+      assertEqual(graph.replicationFactor, undefined);
+      assertEqual(graph.numberOfShards, undefined);
+      assertEqual(graph.smartGraphAttribute, undefined);
```
[the remark this answers is not in the document]
Comment (romanatarango) [reply tests/js/server/dump/dump-test.inc:1424]: It just checks that the keys are not contained in the object graph. graph itself is already there.

File: tests/js/server/dump/dump-test.inc [hunk 5]
```diff
@@ -1281,7 +1377,203 @@
...
+      assertEqual(graph.edgeDefinitions, [
+        {
+          "collection": edges,
+          "from": [verticesSat],
+          "to": [verticesNonSat]
+        }
+      ]);
+      assertEqual(graph.orphanCollections, [orphans]);
+
+      for (const collName of [verticesNonSat, verticesSat, orphans]) {
```
[the remark this answers is not in the document]
Comment (romanatarango) [reply tests/js/server/dump/dump-test.inc:1435]: This was a new variable `properties`, now renamed.

[the remark this answers is not in the document]
Comment (hkernbach) [reply tests/js/server/dump/dump-test.inc]: As I already said: It checked satellite collections and smart collections in one go. They have different properties - so this cannot and was not working. Fixed this now.

File: tests/js/server/dump/dump-setup-common.inc (hunk 1 above)
[the remark this answers is not in the document]
Comment (hkernbach) [reply tests/js/server/dump/dump-setup-common.inc]: I've created another wrapper method so do have them seperated between "real exported methods" and helper only methods. This is fixed now.

File: tests/js/server/dump/dump-setup-common.inc (hunk 3 above)
[the remark this answers is not in the document]
Comment (hkernbach) [reply tests/js/server/dump/dump-setup-common.inc]: I do not see any method called: assertCollectionsAreNew. Anyway, now removed all locations which are trying to drop things as this is an setup error - as we want to dump & restore all collections. We do not want to hide or forget some collections.

File: tests/js/server/dump/dump-setup-common.inc [hunk 6]
```diff
@@ -481,49 +562,338 @@ function createSmartGraphSingleServer() {
...
+
+  const hybridGraphName = "UnitTestDumpHybridSmartGraphSingleServer";
+  const verticesNonSat = "UnitTestDumpHybridSmartVerticesNonSatSingleServer";
+  const verticesSat = "UnitTestDumpHybridSmartVerticesSatSingleServer";
+  const edges = "UnitTestDumpHybridSmartEdgesSingleServer";
+  const orphans = "UnitTestDumpHybridSmartOrphansSingleServer";
+
+  if (sgm._exists(hybridGraphName)) {
+    sgm._drop(hybridGraphName, true);
+  }
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply tests/js/server/dump/dump-setup-common.inc]: Fixed this.

File: tests/js/server/dump/dump-setup-common.inc [hunk 7]
```diff
@@ -481,49 +562,338 @@ function createSmartGraphSingleServer() {
...
+  const orphans = "UnitTestDumpHybridSDisjointmartOrphansSingleServer";
+
+  if (sgm._exists(hybridGraphName)) {
+    sgm._drop(hybridGraphName, true);
+  }
+  db._drop(edgesSatA);
+  db._drop(edgesAB);
+  db._drop(verticesANonSat);
+  db._drop(verticesSat);
+  db._drop(orphans);
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply tests/js/server/dump/dump-setup-common.inc]: Fixed this.

File: tests/js/server/dump/dump-setup-common.inc [hunk 8]
```diff
@@ -697,13 +1068,29 @@ function createSmartGraphRegressionTest() {
...
+
 exports.createSmartGraphSingleServer = createSmartGraphSingleServer;
+exports.createSatelliteGraphSingleServer = createSatelliteGraphSingleServer;
+exports.createDisjointGraphSingleServer = createDisjointGraphSingleServer;
+exports.createHybridSmartGraphSingleServer = createHybridSmartGraphSingleServer;
+exports.createHybridDisjointSmartGraphSingleServer = createHybridDisjointSmartGraphSingleServer;
+// todo: hybrid disjoint
+// A -> B -> C
+// A smart, B satellite, C smart: A-B-C okay,
+// A-C okay only if the same attribute
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply tests/js/server/dump/dump-setup-common.inc]: Removed that todo.

File: tests/js/server/dump/dump-setup-common.inc [hunk 9]
```diff
@@ -481,49 +562,338 @@ function createSmartGraphSingleServer() {
...
+  // drop the graph and collections (if any)
+  if (graphModule._exists(names.smartGraphName)) {
+    graphModule._drop(names.smartGraphName, true);
+  }
+  db._drop(names.edges);
+  db._drop(names.fromVertices);
+  if (names.toVertices !== names.fromVertices) {
+    db._drop(names.toVertices);
+  }
+  db._drop(names.orphans);
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply tests/js/server/dump/dump-setup-common.inc]: I've fixed this now.

File: tests/js/server/dump/dump-test.inc [hunk 10]
```diff
@@ -694,24 +738,24 @@
...
+      //   let analyzer = analyzers.analyzer("custom");
+      //   assertEqual(db._name() + "::custom", analyzer.name());
+      //   assertEqual("delimiter", analyzer.type());
+      //   assertEqual(Object.keys(analyzer.properties()).length, 1);
+      //   assertEqual(" ", analyzer.properties().delimiter);
+      //   assertEqual(1, analyzer.features().length);
+      //   assertEqual("frequency", analyzer.features()[0]);
+      //
+      //   assertNull(analyzers.analyzer("custom_dst"));
+      // },
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply tests/js/server/dump/dump-test.inc]: I've re-enabled them now.

File: tests/js/server/dump/dump-test.inc [hunk 11]
```diff
@@ -1281,7 +1377,203 @@
...
+        assertEqual(properties.numberOfShards, 1);
+        assertEqual(properties.shardingStrategy, 'enterprise-compat');
+      }
+      assertEqual(db._collection(verticesSat).properties().replicationFactor, 1);
+      assertEqual(db._collection(verticesNonSat).properties().replicationFactor, 1);
+
+      const eColl = db._collection(edges);
+      const properties = eColl.properties();
+
+      assertFalse(properties.isSmart);
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply tests/js/server/dump/dump-test.inc]: Fixed this

File: tests/js/server/dump/dump-test.inc [hunk 12]
```diff
@@ -1252,8 +1319,9 @@
         assertTrue(!!g.exists(smartGraphName));
 
         const graph = g.document(smartGraphName);
-        assertCommon(graph);
-        assertCollectionProperties(graph, edgeCollName, vertexCollName, orphanCollName);
+        assertFalse(graph.isDisjoint);
+        assertCommonSmartGraphProperties(graph);
+        assertSmartCollectionProperties(graph, edgeCollName, vertexCollName, orphanCollName, false); // todo
```
[the remark this answers is not in the document]
Comment (hkernbach) [reply tests/js/server/dump/dump-test.inc:1323]: removed unclear todo.

**Comment by hkernbach** [comment]
Jenkins started now. Will merge as soon as jenkins reports green.

**Comment by hkernbach** [comment]
Jenkins green ,will merge now back to feature base branch

**Pull request merged**